Add "Constant" osu!mania mod

chev2 opened the pull request:
Resolves #9802 
This adds a new mod to osu!mania that makes all the slider velocities/speed changes the default value (1).
It only affects difficulty points' speed multiplier as of now. BPM timing points are left unchanged, which may cause issues stated in #10267 (speeding up BPM, and slowing down scroll velocity).
The mod is set to unranked.

chev2 committed:
Add Constant osu!mania mod

bdach commented on osu.Game.Rulesets.Mania/Mods/ManiaModConstant.cs:
```diff
@@ -0,0 +1,32 @@
+
+            foreach (var difficultyPoint in difficultyPoints)
+            {
+                // Set this difficulty point's speed multiplier to 1
+                difficultyPoint.SpeedMultiplier = 1;
```
So what's the plan regarding BPM changes here? I don't think generating a correct set of difficulty control points that would compensate BPM would be overly difficult, so why is this not done? Are you expecting turning off BPM scaling to become a game-wide feature?

chev2 replied:
The compensation for BPM didn't cross my mind, I'll look into that. I do not expect BPM scaling to change. The main issue I was facing was that if I were to reset all timing points as well, there would be the chance that an actual change in BPM on the map would be reset and make the rest of the map off-sync.

bdach replied:
I think adding a difficulty control point for each timing control point, with multiplier equal to (initial BPM / current BPM) should do the trick. In fact I think that is already how constant scrolling speed is mapped? And yeah, just nuking timing points isn't going to work.

abraker95 commented on osu.Game.Rulesets.Mania/Mods/ManiaModConstant.cs:
```diff
@@ -0,0 +1,32 @@
+    public class ManiaModConstant : Mod, IApplicableToBeatmap
+    {
+        public override string Name => "Constant";
+        public override string Acronym => "CT";
```
Change name to "No SV" and change the acronym to "NSV". NSV is better recognized by mania players as an indication in versions of maps without SV.

bdach replied:
Don't think we do three letter acronyms anywhere. Probably won't fit in button?

Joehuu replied:
https://github.com/ppy/osu/blob/575a00dd7deafdc6a9194dec59d79737db44e0ad/osu.Game.Rulesets.Mania/Mods/ManiaModKey10.cs#L10

bdach replied:
Fair enough on that point - still not sure about using SV as an acronym directly in the name. Sort of inaccessible for newcomers (mania doesn't even really have "sliders").

abraker95 replied:
I like to thing mania "SV" stands for scroll velocity, but any mania player knows what SV means because it's hard not to come across the term somewhere.

The wiki details it [here](https://osu.ppy.sh/wiki/en/Guides/osu%21mania_Mapping_Guide#slider-velocity-(sv)). [A pinned thread in the osu!mania subforum](https://osu.ppy.sh/community/forums/topics/155780) uses the term for describing how to map SV. [Countless of mappers use "NSV"](https://osu.ppy.sh/beatmapsets?m=3&q=nsv&s=graveyard) in the difficulty name to indicate it's the same map without SV.

bdach replied:
I still think that "SV" as two letters should not be in the name, regardless of if it's established jargon. I'd be fine with "Constant Scroll" or something, as long as there are actual words there and not two-letter acronyms.

chev2 replied:
I'm fine with "Constant Scroll" too, since gives a better idea as to what this mod does compared to just "Constant". "Constant Scrollspeed" might also work, but it might also be a bit long for a name.
Either way, I think the acronym "CT" should stay the same. "CS" is already an acronym for circle size, but that's not necessarily a mania thing, so I don't know.

peppy replied:
how about FS (fixed scroll)? if we're looking to avoid acronym clashes and using confusing terms like SV.

chev2 commented on osu.Game.Rulesets.Mania/Mods/ManiaModConstant.cs:
```diff
@@ -0,0 +1,32 @@
+        public override string Acronym => "CT";
+        public override double ScoreMultiplier => 1;
+        public override string Description => "No more tricky note speed changes!";
+        public override IconUsage? Icon => FontAwesome.Solid.ArrowDown;
```
I think `FontAwesome.Solid.BalanceScale` looks the best, I'll change it to that.

chev2 committed:
Change mod icon to a better-suited one, add timing point counteraction

chev2 commented on osu.Game.Rulesets.Mania/Mods/ManiaModConstant.cs:
```diff
@@ -0,0 +1,50 @@
+                DifficultyControlPoint diffControlPoint = new DifficultyControlPoint
+                {
+                    SpeedMultiplier = initialBPM / timingPoint.BPM
+                };
```
Regarding this, since difficulty points have a capped speed multiplier value of 0.1-10, there will be cases where the calculated `initialBPM / timingPoint.BPM` exceeds this range. One beatmap that has this is [this one](https://osu.ppy.sh/beatmapsets/586848#mania/1242696). Any thoughts to work around this?

chev2 committed:
Modification by group, remove difficulty speed cap

For every group with a timing point, an associated difficulty point is modified/added. For every group with no timing point but a difficulty point, remove the difficulty point. The second case comes in use as 'Giving every difficulty point a speed multipler of 1' would interfere with difficulty points that counteract timing points.

bdach commented on osu.Game.Rulesets.Mania/Mods/ManiaModConstant.cs:
```diff
@@ -20,30 +21,38 @@ public class ManiaModConstant : Mod, IApplicableToBeatmap
+
+                    DifficultyControlPoint diffControlPoint = new DifficultyControlPoint();
+
+                    diffControlPoint.SpeedMultiplierBindable.MinValue = double.MinValue; // Uncapped minimum value
```
Yeah I'm not sure this should be allowed to be negative (or even zero). I'd agree with

```suggestion
                    diffControlPoint.SpeedMultiplierBindable.MinValue = Precision.DOUBLE_EPSILON; // Uncapped minimum value
```

chev2 replied:
Alright, I'll fix this as well as the precision.

bdach commented on osu.Game.Rulesets.Mania/Mods/ManiaModConstant.cs:
```diff
@@ -20,30 +21,38 @@ public class ManiaModConstant : Mod, IApplicableToBeatmap
+
+                    diffControlPoint.SpeedMultiplierBindable.MinValue = double.MinValue; // Uncapped minimum value
+                    diffControlPoint.SpeedMultiplierBindable.MaxValue = double.MaxValue; // Uncapped maximum value
+                    diffControlPoint.SpeedMultiplierBindable.Precision = 0.000000001d; // Arbitrary value, may need changing
```
Similarly

```suggestion
                    diffControlPoint.SpeedMultiplierBindable.Precision = Precision.DOUBLE_EPSILON;
```

bdach commented on osu.Game.Rulesets.Mania/Mods/ManiaModConstant.cs:
```diff
@@ -20,30 +21,38 @@ public class ManiaModConstant : Mod, IApplicableToBeatmap
+                    group.Add(diffControlPoint);
+                }
+                // Else, there is no timing control point, but a difficulty control point
+                else if (difficultyControlPoints.Any())
```
Not sure this is correct - I don't think there's anything stopping a group from having a timing point _and_ a difficulty control point? Should just probably be

```suggestion
                if (difficultyControlPoints.Any())
```

chev2 replied:
If it were to check for a difficulty point after it checks for a timing point, there's a possibility that it will remove the difficulty point, which will undo the timing point's counteraction. From my understanding, at least. That's why it's an `else if`. If the group has both a timing and difficulty point in the first place, the difficulty point will be overridden anyway.

bdach replied:
Oh, because there's a difficulty point potentially added before... Yeah I think that's a bit backwards. Should clean up the group from any difficulty points first before proceeding to add own overrides here I reckon.

chev2 committed:
Change speed multiplier min value & precision

chev2 committed:
Check for difficulty points before timing points

bdach commented:
On re-testing this, there seem to be blocking issues:

* Hold note duration does not look to change properly on maps with changing BPM. This is most likely due to changing the control points too late in the parsing/conversion process? Not sure right now.
* Worse yet, if the mod is activated and then de-activated, the control points won't get restored to their previous state. This will be due to internal caching of the beatmap and not restoring the changes to control points which were made.

The second point is the bigger issue and I believe this is also an issue that the editor has, and will need addressing properly separately.

chev2 committed:
Change mod name & acronym

chev2 committed:
Change mod class & file name, as well

bdach commented:
So the timing point pollution issue is fixed, and I can believe that it was causing the mis-timings I was seeing (because the mod could apply itself to a beatmap multiple times), but there's still one issue left - namely that this mod doesn't work on converts, because of this:

https://github.com/ppy/osu/blob/dd809df07654bb0b8ba104b6f078c874272aa752/osu.Game.Rulesets.Mania/UI/DrawableManiaRuleset.cs#L78-L80

Don't really have any smart ideas how to counteract this, save for just publicising this `isForCurrentRuleset` field...

chev2 committed:
Update to allow compatibility for converts

bdach commented:
@smoogipoo when you have the time, I'd like to know whether you're fine with the [workaround proposed above](https://github.com/ppy/osu/pull/11354#issuecomment-758225968) which was applied in fe92083 by @chev2, or if you have a better idea on how to make this work.

smoogipoo commented:
Couldn't this PR be implemented by using the `ConstantScrollAlgorithm`? It would remove the mess of adding "counteracting" control points and the preserve bool.

I'd say find a way to make the scroll algorithm of `DrawableScrollingRuleset` editable.

bdach commented:
I'd looked into that but ditched the effort due to reasons I don't currently remember. I think I couldn't figure out a way to get the drawable hierarchy to switch from one scrolling mode to another, due to it being cached and all that.

Would have probably warranted a separate pull and I don't even know if we have the bandwidth to look at that right now.

smoogipoo commented:
I think `IApplicableToDrawableRuleset` should be usable. As long as it's done before the first update frame, SHOC should handle fine.

bdach commented:
Making the algorithm mutable itself was the fiddly part. I might look into this again but no promises.

peppy commented:
maybe worth getting this one merged and converting it to use a scroll algorithm as a second step? the implementation is quite isolated so i don't see much harm in it.

smoogipoo commented:
Do note that if this is merged, your timing point cloning PR needs changes.

peppy commented:
Even without the mod applied? Maybe I'm misunderstanding the gravity of the ruleset-side change.

smoogipoo commented:
Yes. Bottom line is this mod modifies the beatmap's `ControlPointInfo`. It needs a mod interface similar to `IApplicableToBeatmapDifficulty` that clones it to be safe.

peppy commented on osu.Game.Rulesets.Mania/UI/DrawableManiaRuleset.cs:
```diff
@@ -55,6 +55,11 @@ public class DrawableManiaRuleset : DrawableScrollingRuleset<ManiaHitObject>
+        /// <summary>
+        /// Whether difficulty points should be taken into account, even on convert maps.
+        /// </summary>
+        public bool PreserveDifficultyPoints { get; set; }
```
Re-thinking this one, i really don't think this should exist..

bdach replied:
I agree and was reluctant to suggest it in the first place, but in the grand scheme of things I don't deem this pull important enough to spend time on doing things properly by substituting the scroll algorithm.
